### PR TITLE
Allow snapshots to be reused in queries

### PR DESCRIPTION
**Pull request by martinsumner**
Allow for a full bookie snapshot to be re-used for multiple queries, not just KV fetches.

### CODE REVIEW

**Comment by martinsumner** [comment]
https://github.com/martinsumner/leveled/issues/413

File: src/leveled_bookie.erl [hunk 1]
```diff
@@ -1849,22 +1886,43 @@ set_options(Opts, Monitor) ->
...
+%% runners post-query cleanup action
+%% 
+%% When the bookie is a snapshot, a fresh snapshot should not be taken, the
+%% previous snapshot should be used instead.  Also the snapshot should not be
+%% closed as part of the post-query activity as the snapshot may be reused, and
+%% should be manually closed.
 return_snapfun(State, SnapType, Query, LongRunning, SnapPreFold) ->
-    case SnapPreFold of
-        true ->
+    CloseFun =
```
Comment (ThomasArts) [inline src/leveled_bookie.erl:1912]: Sanity check:
Following flow it can be seen that `pcl_close` always returns `ok` if it returns... which means that `ink_close` will be called eventually. But also ink_close will always return `ok` if it returns (there is no corresponding match in line 1918)

Alternative is to check for Pid right away when function is created:
```
fun(LS0, JS0) when is_pid(JS0) ->
    fun() -> ok = leveled_penciller:pcl_close(LS0), leveled_inker:ink_close(JS0) end;
    (LS0, _) ->
       fun() -> leveled_penciller:pcl_close(LS0) end
   end
```

File: src/leveled_bookie.erl
Comment (ThomasArts) [inline src/leveled_bookie.erl:1]: Not completely clear to me whether the extra logging, via `leveled_monitor:log_add(Monitor, ForcedLogs)` is cleaned up when the process dies. There is a DOWN handling that does not call log_remove, but probably that is handled elsewhere anyway.

File: src/leveled_penciller.erl [hunk 2]
```diff
@@ -654,9 +654,12 @@ init([LogOpts, PCLopts]) ->
...
                                                 LongRunning),
             leveled_log:log(p0001, [self()]),
-            {ok, State#state{is_snapshot = true,
-			     bookie_monref = BookieMonitor,
-			     source_penciller = SrcPenciller}};
+            {ok,
+                State#state{
+                    is_snapshot = true,
+                    clerk = undefined,
+			        bookie_monref = BookieMonitor,
```
Comment (ThomasArts) [inline src/leveled_penciller.erl]: Ugly tabs in code
```suggestion
                    bookie_monref = BookieMonitor,
```

File: src/leveled_penciller.erl [hunk 3]
```diff
@@ -654,9 +654,12 @@ init([LogOpts, PCLopts]) ->
...
+			        source_penciller = SrcPenciller}};
```
Comment (ThomasArts) [inline src/leveled_penciller.erl]:
```suggestion
                    source_penciller = SrcPenciller}};
```

File: src/leveled_penciller.erl [hunk 4]
```diff
@@ -1150,22 +1153,33 @@ handle_cast({fetch_levelzero, Slot, ReturnFun}, State) ->
     ReturnFun(lists:nth(Slot, State#state.levelzero_cache)),
     {noreply, State};
 handle_cast({log_level, LogLevel}, State) ->
-    PC = State#state.clerk,
-    ok = leveled_pclerk:clerk_loglevel(PC, LogLevel),
-    ok = leveled_log:set_loglevel(LogLevel),
+    case State#state.clerk of
```
Comment (ThomasArts) [inline src/leveled_penciller.erl]: This code is repeated so often that it almost feels worth to add it to leveled_utils as
```
log_if_pid(undefined, _) -> ok;
log_if_pid(Pid, Fun) when is_pid(Pid) -> Fun().
```
And call it in different forms as:
`log_if_pid(State#state.clerk, fun()  -> leveled_pclerk:clerk_loglevel(PC, LogLevel) end)`

Matter of taste

File: src/leveled_runner.erl [hunk 5]
```diff
@@ -301,13 +275,7 @@ tictactree(SnapFun, {Tag, Bucket, Query}, JournalCheck, TreeSize, Filter) ->
 
             % Close down snapshot when complete so as not to hold removed
             % files open
-            ok = leveled_penciller:pcl_close(LedgerSnap),
-            case JournalCheck of
```
Comment (ThomasArts) [inline src/leveled_runner.erl:305]: Now you build the afterfunction as part of the SnapFun.
But it is hard to follow from the code whether indeed you do this JournalCheck or in line 608 DefferedFetch check.
The code gets clearer, but difficult to validate that it has the same behaviour.
In leveled_bookie the return_snap_fun seems to always close both files, whereas code here suggests that that need not always be the case.

File: src/leveled_runner.erl [hunk 6]
```diff
@@ -602,16 +556,6 @@ foldobjects(SnapFun, Tag, KeyRanges, FoldObjFun, DeferredFetch,
                                                                 LastModRange,
                                                                 LimitByCount)
                 end,
-            AfterFun = 
-                fun() ->
-                    ok = leveled_penciller:pcl_close(LedgerSnapshot),
-                    case DeferredFetch of 
-                        {true, false} ->
```
Comment (ThomasArts) [inline src/leveled_runner.erl:609]: Here we actively did not close the JournalSnapshot.. but is that still the case in the refactored code?

File: src/leveled_codec.erl
Comment (ThomasArts) [inline src/leveled_codec.erl:1]: This seems rather unrelated to the rest of the refactoring?

Comment (martinsumner) [reply src/leveled_codec.erl:1]: This came up when using this patch in riak_kv, it is related to noise reduction in logs.  It isn't normally expected for a tag not to be in the strategy, and so ic012 is `WARN` level log - generally this logs existed because we expect that there may have been some sort of key corruption.  However, when we set a dummy tag because there was a MPUT, that is expected.

The change stops the log noise in this case, but also makes sure via the code coverage checks that we are checking a genuinely invalid tag in the tests

Comment (martinsumner) [reply src/leveled_codec.erl:1]: I will add a comment, as it took me a long time to work out why I'd done this, or maybe re-organise the code to make this more obvious

File: src/leveled_bookie.erl
Comment (martinsumner) [reply src/leveled_bookie.erl:1]: The monitor and logging is specific to the instance of leveled, and dies with the instance.  The snapshots don't have a monitor process (so stats aren't gathered from snapshots).  the change for this PR is to make sure that when we add_logs to the monitor (to tell it to override the default log level for a given log reference), if we do this to a snapshot it doesn't add to the monitor which doesn't exist.

File: src/leveled_runner.erl (hunk 5 above)
Comment (martinsumner) [reply src/leveled_runner.erl:305]: The tests are comprehensive on this I think, so I'm happy with the risks of the change.  

The only reason for having a JournalSnapshot was to do the JournalCheck.  Previously this was a bit specific ... as a developer I know that for this runner the JournalSnapshot will only present when JournalCheck is true - so test for JournalCheck.  Now with AfterFun/SnapFun it is cleaner that only the Bookie, not the Runner, needs knowledge of whether a snapshot was taken (and therefore if it should be closed).

File: src/leveled_runner.erl (hunk 6 above)
Comment (martinsumner) [reply src/leveled_runner.erl:609]: Yes, as above.  Before we were looking for specific flags to recall if a Journal snapshot would have been taken,  But now, as we create the AfterFun when we create the SnapFun there is no need for function-specific knowledge downstream.

File: src/leveled_bookie.erl (hunk 1 above)
Comment (martinsumner) [reply src/leveled_bookie.erl:1912]: I think I'm happy with it as-is.  Double-checked as well `ok` is the only return.

File: src/leveled_penciller.erl (hunk 2 above)
Comment (martinsumner) [reply src/leveled_penciller.erl]: Fixed

File: src/leveled_penciller.erl (hunk 3 above)
Comment (martinsumner) [reply src/leveled_penciller.erl]: Fixed

File: src/leveled_penciller.erl (hunk 4 above)
Comment (martinsumner) [reply src/leveled_penciller.erl]: Repeated code now in function

**Comment by martinsumner** [comment]
Reviewed https://github.com/martinsumner/leveled/pull/415
